Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+2. GC interactions with hidden return buffer. The convention for whether the hidden return buffer can be allocated in the GC heap, and thus needs to be written to using write barrier.
+3. 
+These conventions would be codified as well.
+ 
+### Performance Impact
+
+Because it was already the case that methods outside the current module had to use an indirect call, versionability does not introduce more overhead for non-virtual method calls if inlinin was not done. Thus the main cost of having making the native code version resilient is the requirement that no cross version bubble inlining can happen.
+ 
+The best solution to this problem is to avoid 'chatty' library designs (Unfortunately, `IEnumerable`, is such a chatty design, where each iteration does a `MoveNext` and `Current` property fetch). Another mitigation is the one mentioned previously: to allow clients of the library to selective JIT compile some methods that make these chatty calls. Finally you can also use new custom `NonVersionableAttribute` attribute, which effectively changes the versioning contract to indicate that the library supplier has given up his right to change that method's body and thus it would be legal to inline.
+ 
+The proposal is to disallow cross-version bubble inlining by default, and selective allow inlining for critical methods (by giving up the right to change the method).
+
+Experiment with disabled cross-module inlining with the selectively enabled inlining of critical methods showed no visible regression in ASP.NET throughput.
+
+## Non-Virtual calls as the baseline solution to all other versioning issues
+
+It is important to observe that once you have mechanism for doing non-virtual function calls in a version resilient way (by having an indirect CALL through a slot that that can be fixed lazily at runtime, all other versioning problems _can_ be solved in that way by calling back to the 'definer' module, and having the operation occur there instead. Issues associated with this technique
+
+1. You will pay the cost of a true indirection function call and return, as well as any argument setup cost. This cost may be visible in constructs that do not contain a call naturally, like fetching string literals or other constants. You may be able to get better performance from another technique (for example, we did so with instance field access).
+2. It introduces a lot of indirect calls. It is not friendly to systems that disallow on the fly code generation. A small helper stub has to be created at runtime in the most straightforward implementation, or there has to be a scheme how to pre-create or recycle the stubs.
+3. It requires that the defining assembly 'know' the operations that it is responsible for defining. In general this could be fixed by JIT compiling whatever is needed at runtime (which the needed operations are known), but JIT compiling is the kind of expensive operations that we are trying to avoid at runtime.
```

Comment:
~~which the needed operations are known~~ *where* the needed operations are known